File: server/Services/FavoritesService.cs



namespace all_spice_dotnet.Services;

public class FavoritesService
{
  public FavoritesService(FavoritesRepository repository)
  {
    _repository = repository;
  }
  private readonly FavoritesRepository _repository;

  internal FavoriteRecipe CreateFavorite(Favorite favoriteData)
  {
    FavoriteRecipe favoriteRecipe = _repository.CreateFavorite(favoriteData);
    return favoriteRecipe;
  }

  internal List<FavoriteRecipe> GetAccountFavoriteRecipes(string userId)
  {
    List<FavoriteRecipe> favoriteRecipes = _repository.GetAccountFavoriteRecipes(userId);
    return favoriteRecipes;
  }

  public Favorite GetFavoriteById(int favoriteId)
  {
    Favorite favorite = _repository.GetFavoriteById(favoriteId);

    if (favorite == null) throw new Exception("Invalid favorite id");

    return favorite;
  }

  internal string DeleteFavorite(int favoriteId, string userId)
  {
    Favorite favorite = GetFavoriteById(favoriteId);

    if (favorite.AccountId != userId) throw new Exception("You can not un-favorite some else's recipe!");

    _repository.DeleteFavorite(favoriteId);

    return "No longer a favorite!";
  }
}

File: client/src/services/FavoritesService.js
import { logger } from "@/utils/Logger.js"
import { api } from "./AxiosService.js"
import { FavoriteRecipe } from "@/models/Favorite.js"
import { AppState } from "@/AppState.js"

class FavoritesService{
  async createFavorite(favoriteData) {
    const response = await api.post('api/favorites', favoriteData)
    logger.log("CREATED FAVORITE", response.data)
    const recipe = new FavoriteRecipe(response.data)
    AppState.favoriteRecipes.push(recipe)
  }
  
  async deleteFavorite(favorite) {
    const response = await api.post(`api/favorites/${favorite}`)
    logger.log("DELETED FAVORITE", response.data)
  }
}

export const favoritesService = new FavoritesService()